_posts/2022-05-19-quinoa-modern-ui-with-no-hassles.adoc

```diff
@@ -0,0 +1,74 @@
+
+=== A bit of context
+
+Alongside Quarkus, interacting with web frameworks such as React and Angular, is a natural need. Many users have been requesting a smooth integration since the early stages of Quarkus (https://github.com/quarkusio/quarkus/issues/588[#588]). This is why I created the Quinoa extension for Quarkus.
+
+Since then (3 years), link:#existing-articles[many blog articles] have popped up to help in that matter, most of them taking advantage of the frontend-maven-plugin. As creator of code.quarkus.io, I have also been confronted with this situation and I, like many others, have filled the gap using this plugin. It does the job for sure and has been good enough to fill the need for the time being. Unfortunately the overall process could be more streamlined and easier to get started with, it involves a lot of configuration, some tricky setup to copy the resources to the right place and test the frontend at the right moment. Live coding requires starting two processes and a proxy to make it work together.
+
+=== Let's start cooking
+
+With this experience behind me, I was in a good position to contribute something to ease our lives a bit. I put an eye to all the needs, issues, propositions and blog articles and came up with a new extension to create healthy modern user-interfaces (UI) with Quarkus: Quinoa (standing for Quarkus UI with NO hAssles :)
+
+Quinoa is taking care of all the Web UI build/wiring/live-config hassles to let you take care of your web application logic. Either to create a full stack Single Page Application or Microservice (micro-frontend); things are made easier for you.
+
+NodeJS has great tooling to create modern frontend web applications. Whilst Quarkus has a broad ecosystem to build reliable modern backends in the cloud (microservices, rest, DBs, events, cloud native…). Quinoa is the glue to get the best of both worlds.
+
+The only prerequisite for the NodeJS project to be compatible with Quinoa is to contain the `build` script to build the Web UI and deliver static files in a specific directory. Since teams can be different, to make everyone happy, it’s possible to either locate the UI in the Quarkus project (default is `src/main/webui`) or outside (as soon as it’s available on the disk at build time).
+
+It is not necessary (but possible) to https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#package-manager[configure the package manager] (NPM, Yarn or PNPM), it will be auto-detected depending on the project lockfile.
+
+Then Quinoa will take care of the generated UI static files, they will be bundled in the jar or native binary and efficiently served alongside your backend (by Vert.x). Also something that UI developers were asking was to enable Single Page App routing. It is now just a https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#spa-routing[boolean configuration].
```

Comment:
```suggestion
Quinoa will take care of the generated UI static files. They will be bundled in the jar or native binary and efficiently served alongside your backend (by Vert.x). UI developers also were asking was to enable Single Page App routing. It is now just a https://quarkiverse.github.io/quarkiverse-docs/quarkus-quinoa/dev/#spa-routing[boolean configuration].
```

does the "efficiently" matter here btw. anything specially done elsewhere where it is "ineffeciently" served?